implementations/python/mzlib/index/memory.py

```diff
@@ -13,12 +14,29 @@
 class IndexRecord(object):
     __slots__ = ('number', 'offset', 'name', 'analyte', 'attributes')
 
+    number: int
+    offset: int
+    name: str
+    analyte: Any
+    attributes: Optional[Dict[str, Any]]
+
     def __init__(self, number, offset, name, analyte, attributes=None):
         self.number = number
         self.offset = offset
         self.name = name
         self.analyte = analyte
-        self.attributes = attributes or {}
+        self.attributes = attributes
+
+    def get(self, key: str, default=None) -> Any:
+        if self.attributes:
+            return self.attributes.get(key, default)
+        return default
+
+    def set(self, key: str, value: Any):
+        if self.attributes:
```

Comment:
Better to check for `is None`?

Reply:
I've done this. Ultimately this only changes the behavior when `self.attributes` isn't a plain `dict` with either a different invariant or a different `__bool__` implementation.

In general, I think I need to change `IndexRecord` to implement the full `MutableMapping` interface rather than this "get/set" pair.